File: app/src/main/java/com/jakebarnby/filemanager/models/OneDriveSource.java
package com.jakebarnby.filemanager.models;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.support.v4.app.Fragment;
import android.util.Log;

import com.jakebarnby.filemanager.managers.OneDriveFactory;
import com.jakebarnby.filemanager.models.files.OneDriveFile;
import com.jakebarnby.filemanager.models.files.SourceFile;
import com.jakebarnby.filemanager.util.Constants;
import com.jakebarnby.filemanager.util.TreeNode;
import com.microsoft.graph.concurrency.ICallback;
import com.microsoft.graph.core.ClientException;
import com.microsoft.graph.core.DefaultClientConfig;
import com.microsoft.graph.core.IClientConfig;
import com.microsoft.graph.extensions.DriveItem;
import com.microsoft.graph.extensions.GraphServiceClient;
import com.microsoft.graph.extensions.IDriveItemCollectionPage;
import com.microsoft.identity.client.AuthenticationCallback;
import com.microsoft.identity.client.AuthenticationResult;
import com.microsoft.identity.client.MsalClientException;
import com.microsoft.identity.client.MsalException;
import com.microsoft.identity.client.MsalServiceException;
import com.microsoft.identity.client.MsalUiRequiredException;
import com.microsoft.identity.client.PublicClientApplication;
import com.microsoft.identity.client.User;

import java.util.List;

/**
 * Created by jakebarnby on 2/08/17.
 */

public class OneDriveSource extends Source {

    public static final String      TAG = "ONEDRIVE";
    private final static String     CLIENT_ID = "019d333e-3e7f-4c04-a214-f12602dd5b10";
    private static final String[]   SCOPES = {"https://graph.microsoft.com/Files.ReadWrite"};

    private PublicClientApplication mClient;
    private AuthenticationResult mAuthResult;

    public OneDriveSource(String sourceName, SourceListener listener) {
        super(sourceName, listener);
    }

    @Override
    public void authenticateSource(Context context) {

    }

    public PublicClientApplication getClient() {
        return mClient;
    }

    public void authenticateSource(Fragment fragment) {
        mClient = new PublicClientApplication(fragment.getContext(), CLIENT_ID);
        mSourceListener.onLoadStarted();
        if (!isLoggedIn()) {
            List<User> users;
            try {
                users = mClient.getUsers();
                if (users != null && users.size() == 1) {
                    mClient.acquireTokenSilentAsync(SCOPES, users.get(0), getAuthSilentCallback(fragment.getContext()));
                } else {
                    mClient.acquireToken(fragment, SCOPES, getAuthInteractiveCallback(fragment.getContext()));
                }
            } catch (MsalClientException e) {
                Log.d(TAG, "MSAL Exception Generated while getting users: " + e.toString());
                mSourceListener.onLoadAborted();

            } catch (IndexOutOfBoundsException e) {
                Log.d(TAG, "User at this position does not exist: " + e.toString());
                mSourceListener.onLoadAborted();
            }
        }
    }

    @Override
    public void loadSource(Context context) {
        if (!isFilesLoaded() && mAuthResult != null) {
            if (!checkConnectionActive(context)) return;
            final IClientConfig mClientConfig = DefaultClientConfig
                    .createWithAuthenticationProvider(iHttpRequest -> {
                        iHttpRequest.addHeader("Authorization", String.format("Bearer %s", mAuthResult.getAccessToken()));
                    });

            OneDriveFactory
                    .getInstance()
                    .setGraphClient(new GraphServiceClient
                            .Builder()
                            .fromConfig(mClientConfig)
                            .buildClient());

            OneDriveFactory
                    .getInstance()
                    .getGraphClient()
                    .getMe()
                    .getDrive()
                    .getRoot()
                    .buildRequest()
                    .get(new ICallback<DriveItem>() {
                        @Override
                        public void success(DriveItem driveItem) {
                            new OneDriveFileSystemLoader().executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, driveItem);
                        }

                        @Override
                        public void failure(final ClientException ex) {
                            // Handle failure
                        }
                    });
        }
    }


    /**
     * Check for a valid access token and store it in shared preferences if found, then load the source
     */
    public void checkForAccessToken(Context context) {
        SharedPreferences prefs = context.getSharedPreferences(Constants.SharedPrefs.PREFS, Context.MODE_PRIVATE);
        String accessToken = prefs.getString("onedrive-access-token", null);
        if (mAuthResult != null) {
            if (accessToken == null || !mAuthResult.getAccessToken().equals(accessToken)) {
                accessToken = mAuthResult.getAccessToken();
                prefs.edit().putString("onedrive-access-token", accessToken).apply();
            } else {
                if (!isLoggedIn()) {
                    authenticateSource(context);
                } else {
                    loadSource(context);
                }
            }
        }
        if (accessToken != null) {
            if (!isLoggedIn()) {
                authenticateSource(context);
            } else {
                loadSource(context);
            }
        }
    }


    /**
     * OnSpaceCheckCompleteListener method for acquireTokenSilent calls
     * Looks if tokens are in the cache (refreshes if necessary and if we don't forceRefresh)
     * else errors that we need to do an interactive request.
     *
     * @return
     */
    private AuthenticationCallback getAuthSilentCallback(Context context) {
        return new AuthenticationCallback() {
            @Override
            public void onSuccess(AuthenticationResult authenticationResult) {
                mAuthResult = authenticationResult;
                setLoggedIn(true);
                checkForAccessToken(context);
            }

            @Override
            public void onError(MsalException exception) {
                if (exception instanceof MsalClientException) {
                    mSourceListener.onLoadAborted();
                } else if (exception instanceof MsalServiceException) {
                    mSourceListener.onLoadAborted();
                } else if (exception instanceof MsalUiRequiredException) {
                    mSourceListener.onLoadAborted();
                }
            }

            @Override
            public void onCancel() {
                Log.d(TAG, "User cancelled login.");
            }
        };
    }

    /**
     * OnSpaceCheckCompleteListener used for interactive request.  If succeeds we use the access
     * token to call the Microsoft Graph. Does not check cache
     *
     * @return
     */
    private AuthenticationCallback getAuthInteractiveCallback(Context context) {
        return new AuthenticationCallback() {
            @Override
            public void onSuccess(AuthenticationResult authenticationResult) {
                mAuthResult = authenticationResult;
                setLoggedIn(true);
                checkForAccessToken(context);
            }

            @Override
            public void onError(MsalException exception) {
                if (exception instanceof MsalClientException) {
                    mSourceListener.onLoadAborted();
                } else if (exception instanceof MsalServiceException) {
                    mSourceListener.onLoadAborted();
                }
            }

            @Override
            public void onCancel() {
                Log.d(TAG, "User cancelled login.");
            }
        };
    }

    private class OneDriveFileSystemLoader extends AsyncTask<DriveItem, Void, TreeNode<SourceFile>> {

        private TreeNode<SourceFile> rootFileTreeNode;
        private TreeNode<SourceFile> currentLevelNode;

        @Override
        protected TreeNode<SourceFile> doInBackground(DriveItem... driveItems) {
            SourceFile rootSourceFile = new OneDriveFile(driveItems[0]);
            rootSourceFile.setDirectory(true);
            rootFileTreeNode = new TreeNode<>(rootSourceFile);
            currentLevelNode = rootFileTreeNode;
            setCurrentDirectory(rootFileTreeNode);
            setQuotaInfo(OneDriveFactory.getInstance().getStorageStats());

            return parseFileTree(driveItems[0]);
        }

        /**
         * @param currentDirectory
         */
        private TreeNode<SourceFile> parseFileTree(DriveItem currentDirectory) {
            IDriveItemCollectionPage items = OneDriveFactory
                    .getInstance()
                    .getGraphClient()
                    .getMe()
                    .getDrive()
                    .getItems(currentDirectory.id)
                    .getChildren()
                    .buildRequest()
                    .select("id,name,webUrl,folder,size,createdDateTime,lastModifiedDateTime")
                    .expand("thumbnails")
                    .get();

            List<DriveItem> pageItems = items.getCurrentPage();
            long dirSize = 0L;
            if (pageItems != null) {
                for (DriveItem file : pageItems) {
                    SourceFile sourceFile = new OneDriveFile(file);
                    if (file.folder != null) {
                        currentLevelNode.addChild(sourceFile);
                        currentLevelNode = currentLevelNode.getChildren().get(currentLevelNode.getChildren().size() - 1);
                        parseFileTree(file);
                        currentLevelNode.getParent().getData().addSize(currentLevelNode.getData().getSize());
                        currentLevelNode = currentLevelNode.getParent();
                    } else {
                        dirSize += file.size;
                        currentLevelNode.addChild(sourceFile);
                    }
                }
                currentLevelNode.getData().addSize(dirSize);
            }
            return rootFileTreeNode;
        }

        @Override
        protected void onPostExecute(TreeNode<SourceFile> fileTree) {
            super.onPostExecute(fileTree);
            TreeNode.sortTree(fileTree, (node1, node2) -> {
                int result = Boolean.valueOf(!node1.getData().isDirectory()).compareTo(!node2.getData().isDirectory());
                if (result == 0) {
                    result = node1.getData().getName().toLowerCase().compareTo(node2.getData().getName().toLowerCase());
                }
                return result;
            });
            setFilesLoaded(true);
            mSourceListener.onLoadComplete(fileTree);
        }
    }
}

File: app/src/main/java/com/jakebarnby/filemanager/models/sources/onedrive/OneDriveSource.kt
package com.jakebarnby.filemanager.models.sources.onedrive

import android.content.Context
import androidx.fragment.app.Fragment
import com.google.firebase.analytics.FirebaseAnalytics
import com.jakebarnby.filemanager.managers.PreferenceManager
import com.jakebarnby.filemanager.models.Source
import com.jakebarnby.filemanager.models.SourceConnectionType
import com.jakebarnby.filemanager.models.SourceType
import com.jakebarnby.filemanager.models.StorageInfo
import com.jakebarnby.filemanager.util.Constants
import com.jakebarnby.filemanager.util.Constants.Prefs
import com.jakebarnby.filemanager.util.Logger
import com.jakebarnby.filemanager.util.Utils
import com.jakebarnby.filemanager.workers.OneDriveFileTreeWalker
import com.microsoft.graph.core.ClientException
import com.microsoft.graph.core.DefaultClientConfig
import com.microsoft.graph.extensions.*
import com.microsoft.graph.http.GraphServiceException
import com.microsoft.identity.client.*
import com.microsoft.identity.client.exception.MsalException
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import javax.inject.Inject

/**
 * Created by jakebarnby on 2/08/17.
 */
class OneDriveSource @Inject constructor(
    prefsManager: PreferenceManager
) : Source<
    OneDriveCreateFolderParams,
    OneDriveDownloadParams,
    OneDriveUploadParams,
    OneDriveRenameParams,
    OneDriveDeleteParams,
    OneDriveFileTreeWalker,
    DriveItem,
    DriveItem
    >(
    SourceConnectionType.REMOTE,
    SourceType.ONEDRIVE.id,
    prefsManager
) {

    companion object {
        private val SCOPES = arrayOf("https://graph.microsoft.com/Files.ReadWrite")
    }

    @Inject
    lateinit var authClient: IMultipleAccountPublicClientApplication

    @Inject
    lateinit var client: IGraphServiceClient

    var authResult: IAuthenticationResult? = null
    var currentAccount: IAccount? = null

    override val storageInfo: StorageInfo?
        get() {
            try {
                val quota = client
                    .me
                    .drive
                    .buildRequest()
                    .select("quota")
                    .get()
                    .quota

                return StorageInfo(
                    quota?.total ?: 0,
                    quota?.used ?: 0
                )
            } catch (e: GraphServiceException) {
                e.printStackTrace()
            }
            return null
        }

    override suspend fun authenticate(context: Context) {}

    fun authenticate(fragment: Fragment) {
        if (isLoggedIn) {
            return
        }
        listener.onLoadStarted()
        try {
            val accessToken = prefsManager.getString(Prefs.ONEDRIVE_TOKEN_KEY, null)
            val userId = prefsManager.getString(Prefs.ONEDRIVE_NAME_KEY, null)
            if (accessToken != null && userId != null) {
                authClient.acquireTokenSilentAsync(
                    SCOPES,
                    authClient.getAccount(userId),
                    authClient.configuration.defaultAuthority.authorityURL.toString(),
                    getAuthSilentCallback(fragment)
                )
            } else {
                authClient.acquireToken(AcquireTokenParameters(
                    AcquireTokenParameters.Builder()
                        .withScopes(SCOPES.toMutableList())
                        .withFragment(fragment)
                        .withCallback(getAuthInteractiveCallback(fragment))
                ))
            }
        } catch (e: Exception) {
            e.printStackTrace()
            listener.onLoadError(e.message ?: "")
        }

        val clientConfig = DefaultClientConfig.createWithAuthenticationProvider {
            it.addHeader("Authorization", String.format("Bearer %s", authResult!!.accessToken))
        }

        client = GraphServiceClient.Builder()
            .fromConfig(clientConfig)
            .buildClient()
    }

    override suspend fun logout(context: Context) {
        if (!isLoggedIn) {
            return
        }
        prefsManager.savePref(Prefs.ONEDRIVE_TOKEN_KEY, null as String?)
        prefsManager.savePref(Prefs.ONEDRIVE_NAME_KEY, null as String?)

        isLoggedIn = false
        isFilesLoaded = false
        authResult = null
        listener.onLogout()

        Logger.logFirebaseEvent(
            FirebaseAnalytics.getInstance(context),
            Constants.Analytics.EVENT_LOGOUT_ONEDRIVE
        )
    }

    fun checkForAccessToken(fragment: Fragment) {
        var accessToken = prefsManager.getString(Prefs.ONEDRIVE_TOKEN_KEY, null)
        if (authResult != null) {
            if (accessToken == null || authResult?.accessToken != accessToken) {
                accessToken = authResult?.accessToken
                val userId = authResult?.account?.id
                prefsManager.savePref(Prefs.ONEDRIVE_TOKEN_KEY, accessToken)
                prefsManager.savePref(Prefs.ONEDRIVE_NAME_KEY, userId)
            } else {
                if (!isLoggedIn) {
                    authenticate(fragment)
                } else {
                    loadFiles<OneDriveFileTreeWalker>(fragment.context!!)
                }
            }
        }
        if (accessToken != null) {
            if (!isLoggedIn) {
                authenticate(fragment)
            } else {
                loadFiles<OneDriveFileTreeWalker>(fragment.context!!)
                Logger.logFirebaseEvent(
                    FirebaseAnalytics.getInstance(fragment.context!!),
                    Constants.Analytics.EVENT_LOGIN_ONEDRIVE)
            }
        }
    }

    private fun getAuthSilentCallback(fragment: Fragment): AuthenticationCallback =
        object : AuthenticationCallback {
            override fun onSuccess(authenticationResult: IAuthenticationResult) {
                authResult = authenticationResult
                currentAccount = authenticationResult.account
                isLoggedIn = true
                checkForAccessToken(fragment)
            }

            override fun onError(exception: MsalException) {
                listener.onLoadError(exception.errorCode + " " + exception.message)
            }

            override fun onCancel() {
                listener.onLoadAborted()
            }
        }

    private fun getAuthInteractiveCallback(fragment: Fragment): AuthenticationCallback =
        object : AuthenticationCallback {
            override fun onSuccess(authenticationResult: IAuthenticationResult) {
                authResult = authenticationResult
                currentAccount = authenticationResult.account
                isLoggedIn = true
                checkForAccessToken(fragment)
            }

            override fun onError(exception: MsalException) {
                listener.onLoadError(exception.errorCode + " " + exception.message)
            }

            override fun onCancel() {
                listener.onLoadAborted()
            }
        }

    fun getRootFile(): DriveItem? = client.me.drive.root.buildRequest().get()

    fun getFilesByParentId(id: String): List<DriveItem>? {
        val items = mutableListOf<DriveItem>()
        var pages = client.me.drive
            .getItems(id)
            .children
            .buildRequest()
            .select("id,name,webUrl,folder,size,createdDateTime,lastModifiedDateTime")
            .expand("thumbnails")
            .get()

        while (pages != null) {
            items.addAll(pages.currentPage)
            pages = getNextPage(pages)
        }
        return items
    }

    private fun getNextPage(collection: IDriveItemCollectionPage?) =
        collection
            ?.nextPage
            ?.buildRequest()
            ?.select("id,name,webUrl,folder,size,createdDateTime,lastModifiedDateTime")
            ?.expand("thumbnails")
            ?.get()


    override suspend fun createFolder(params: OneDriveCreateFolderParams): DriveItem? {
        val item = DriveItem().apply {
            this.name = params.folderName
            this.folder = Folder()
        }

        return client.me.drive
            .getItems(params.parentId)
            .children
            .buildRequest()
            .post(item)
    }

    override suspend fun download(params: OneDriveDownloadParams): File? {
        val file = File(
            params.destinationPath,
            params.fileName
        )
        try {
            client.me.drive
                .getItems(params.id)
                .content
                .buildRequest()
                .get()
                ?.use {
                    Utils.copyInputStreamToFile(it, file)
                    return file
                }
        } catch (e: IOException) {
            throw e
        } catch (e: ClientException) {
            throw e
        }
        return null
    }

    override suspend fun upload(params: OneDriveUploadParams): DriveItem? {
        val file = File(params.filePath)
        try {
            return withContext(IO) {
                FileInputStream(file).use {
                    val buffer = ByteArray(file.length().toInt())
                    it.read(buffer)

                    client.me.drive
                        .getItems(params.parentId)
                        .getChildren(transformFileName(params.fileName))
                        .content
                        .buildRequest()
                        .put(buffer)
                }
            }
        } catch (e: IOException) {
            throw e
        }
    }

    override suspend fun delete(params: OneDriveDeleteParams) {
        client.me.drive
            .getItems(params.driveId)
            .buildRequest()
            .delete()
    }

    override suspend fun rename(params: OneDriveRenameParams): DriveItem? {
        val item = DriveItem().apply {
            name = params.newName
        }
        return client.me.drive
            .getItems(params.driveId)
            .buildRequest()
            .patch(item)
    }

    private fun transformFileName(fileName: String): String {
        var newName = fileName.replace(Constants.Sources.ONEDRIVE_INVALID_CHARS.toRegex(), "%20")
        if (newName.length > Constants.Sources.MAX_FILENAME_LENGTH) {
            newName = newName.substring(0, Constants.Sources.MAX_FILENAME_LENGTH)
        }
        return newName
    }
}